src/dmclock_server.h

```diff
@@ -1777,8 +1777,14 @@ namespace crimson {
 
 	    l.unlock();
 	    if (!this->finishing) {
-	      typename super::DataGuard g(this->data_mtx);
-	      schedule_request();
+	      auto get_sched_ahead_when = [this] () -> Time {
+	        std::lock_guard<std::mutex> l(sched_ahead_mtx);
+	        return sched_ahead_when;
+	      };
+	      while (get_sched_ahead_when() == TimeZero && this->request_count() > 0) {
+	        typename super::DataGuard g(this->data_mtx);
+	        schedule_request();
+	      }
```

Comment:
i think the first call of `get_sched_ahead_when()` should always return `TimeZero`. so, for better performance, how about rewriting the `while` loop like

```c++
do {
  typename super::DataGuard g(this->data_mtx);
  schedule_request();
} while (get_sched_ahead_when() == TimeZero && has_request());
```

in comparison with the existing behavior, each loop always needs to acquire the `sched_ahead_mtx` lock one more time no matter  if the next request should be scheduled in future or not, i think we can improve the performance further by making `schedule_request()` return a boolean or an enum indicating if the next request should be scheduled right away or not. if it is, we should continue with the loop until hitting a request which will block the `run_sched_ahead()` thread.

Reply:
@tchaikov based on your valuable review, I've modified some lines.